iroh/src/api.rs

```diff
@@ -0,0 +1,104 @@
+use std::collections::{HashMap, HashSet};
+use std::path::Path;
+
+use anyhow::Result;
+use async_trait::async_trait;
+use bytes::Bytes;
+use cid::Cid;
+use libp2p::gossipsub::MessageId;
+use libp2p::{Multiaddr, PeerId};
+
+pub struct Id {
+    pub peer_id: PeerId,
+    pub listen_addrs: Vec<Multiaddr>,
+    pub local_addrs: Vec<Multiaddr>,
+}
+
+#[derive(Debug)]
+pub enum Ping {
+    PeerId(PeerId),
+    Multiaddr(Multiaddr),
+}
+
+pub trait Api<P: P2p, S: Store>: Main + Accessors<P, S> + GetAdd {}
+
+#[async_trait]
+pub trait Main {
+    async fn version(&self) -> Result<String>;
+}
+
+#[async_trait]
+pub trait Accessors<P: P2p, S: Store> {
+    fn p2p(&self) -> Result<P>;
+    fn store(&self) -> Result<S>;
+}
+
+#[async_trait]
+pub trait GetAdd {
+    // XXX get and add are centered around the filesystem.
+    // We can imagine an underlying version that produces a stream of
+    // Out as well.
+    async fn get(&self, cid: &Cid, path: &Path) -> Result<()>;
+    async fn add(&self, path: &Path) -> Result<Cid>;
+}
+
+impl<T: Main + Accessors<P, S> + GetAdd, P: P2p, S: Store> Api<P, S> for T {}
+
+#[async_trait]
+pub trait P2pConnectDisconnect {
+    async fn connect(&self, peer_id: &PeerId, addrs: &[Multiaddr]) -> Result<()>;
+    async fn disconnect(&self, peer_id: &PeerId) -> Result<()>;
+}
+
+// XXX the std::marker::Sync worries me - it's required to be able to use
+// peers from the implementation of peer_ids
+#[async_trait]
+pub trait P2pId: std::marker::Sync {
+    async fn p2p_version(&self) -> Result<String>;
+    async fn local_peer_id(&self) -> Result<PeerId>;
+    async fn addrs_listen(&self) -> Result<Vec<Multiaddr>>;
+    async fn addrs_local(&self) -> Result<Vec<Multiaddr>>;
+    // can be implemented on the trait itself as it combines others
+    async fn id(&self) -> Result<Id>;
+    // async fn addrs gets a map right now
+    async fn peers(&self) -> Result<HashMap<PeerId, Vec<Multiaddr>>>;
```

Comment:
I don't understand this comment. Do you mean it changed from `addrs` to `peers`? I didn't change anything as this API is new, but it may definitely be inconsistent with the existing CLI somewhere. I saw both a `peers` command and an `addr` command in the CLI code if I recall correctly.

I think in general this whole API  needs a rethink the light of the discussion we had yesterday about the complex relationship between all this stuff.

Reply:
yes I did mean from `addrs` to `peers`, my bad. Yea I agree about the API rethink.